deepchem/models/torch_models/infograph.py

```diff
@@ -0,0 +1,435 @@
+    ----------
+    num_features: int
+        Number of node features for each input
+    edge_features: int
+        Number of edge features for each input
+    dim: int
+        Dimension of the embedding
+    """
+
+    def __init__(self, num_features, edge_features, dim):
+        from torch_geometric.nn import NNConv
+        from torch_geometric.nn.aggr import Set2Set
+
+        super(Encoder, self).__init__()
+        self.lin0 = torch.nn.Linear(num_features, dim)
+
+        nn = Sequential(Linear(edge_features, 128), ReLU(),
+                        Linear(128, dim * dim))
+        self.conv = NNConv(dim, dim, nn, aggr='mean', root_weight=False)
+        self.gru = GRU(dim, dim)
+
+        self.set2set = Set2Set(dim, processing_steps=3)
+
+    def forward(self, data):
+        out = F.relu(self.lin0(data.node_features))
+        h = out.unsqueeze(0)
+
+        for i in range(3):
+            m = F.relu(self.conv(out, data.edge_index, data.edge_features))
+            out, h = self.gru(m.unsqueeze(0), h)
+            out = out.squeeze(0)
+            feat_map = out
+
+        out = self.set2set(out, data.graph_index)
+        return out, feat_map
+
+
+class FF(nn.Module):
+    """
+    A feedforward neural network with a skip connection.
+
+    Parameters
+    ----------
+    input_dim: int
+        Dimension of the input
+    dim: int
+        Dimension of the embedding
+    """
+
+    def __init__(self, input_dim, dim):
+        super().__init__()
+        self.block = nn.Sequential(nn.Linear(input_dim, dim), nn.ReLU(),
+                                   nn.Linear(dim, dim), nn.ReLU(),
+                                   nn.Linear(dim, dim), nn.ReLU())
+        self.linear_shortcut = nn.Linear(input_dim, dim)
+
+    def forward(self, x):
+        return self.block(x) + self.linear_shortcut(x)
+
+
+class InfoGraphModel(ModularTorchModel):
+    """
+    Infograph is a semi-supervised graph convolutional network for predicting molecular properties.
+    It aims to maximize the mutual information between the graph-level representation and the
+    representations of substructures of different scales. It does this by producing graph-level
+    encodings and substructure encodings, and then using a discriminator to classify if they
+    are from the same molecule or not.
+
+    References
+    ----------
+    F.-Y. Sun, J. Hoffmann, V. Verma, and J. Tang, “InfoGraph: Unsupervised and Semi-supervised
+    Graph-Level Representation Learning via Mutual Information Maximization.” arXiv, Jan. 17, 2020.
+    http://arxiv.org/abs/1908.01000
+
+    Parameters
+    ----------
+    num_features: int
+        Number of node features for each input
+    edge_features: int
+        Number of edge features for each input
+    dim: int
+        Dimension of the embedding
+    use_unsup_loss: bool
+        Whether to use the unsupervised loss
+    separate_encoder: bool
+        Whether to use a separate encoder for the unsupervised loss
+
+    Examples
```

Comment:
I think this is only the fine tuning use case? Can you add an example for the semisupervised pretraining?

Reply:
This is the semi-supervised training. There is no infograph training loop that is strictly supervised. There is only semi-supervised and unsupervised.

Reply:
Ah I see. Can you add an unsupervised example as well and add a discussion to the doctoring about the two modes of training and how they differ? Lots of detail is good here since this is the first model of its type in deepchem